cmd/precise-code-intel-api-server/internal/api/references.go

```diff
@@ -0,0 +1,344 @@
+		}
+
+		identifier := cursor.Identifier
+		offset := cursor.SkipDumpsWhenBatching
+		limit := s.remoteDumpLimit
+		newOffset := offset
+
+		var packageRefs []db.Reference
+		for len(packageRefs) < limit && newOffset < totalCount {
+			page, err := pager.PageFromOffset(newOffset)
+			if err != nil {
+				return nil, Cursor{}, false, pager.CloseTx(err)
+			}
+
+			if len(page) == 0 {
+				// Shouldn't happen, but just in case of a bug we
+				// don't want this to throw up into an infinite loop.
+				break
+			}
+
+			filtered, scanned := applyBloomFilter(page, identifier, limit-len(packageRefs))
+			packageRefs = append(packageRefs, filtered...)
+			newOffset += scanned
+		}
+
+		var dumpIDs []int
+		for _, ref := range packageRefs {
+			dumpIDs = append(dumpIDs, ref.DumpID)
+		}
+
+		cursor.DumpIDs = dumpIDs
+		cursor.SkipDumpsWhenBatching = newOffset
+		cursor.TotalDumpsWhenBatching = totalCount
+
+		if err := pager.CloseTx(nil); err != nil {
+			return nil, Cursor{}, false, err
+		}
+	}
+
+	for i, batchDumpID := range cursor.DumpIDs {
+		// Skip the remote reference that show up for ourselves - we've already gathered
+		// these in the previous step of the references query.
+		if i < cursor.SkipDumpsInBatch || batchDumpID == dumpID {
+			continue
+		}
+
+		dump, exists, err := s.db.GetDumpByID(context.Background(), batchDumpID)
+		if err != nil {
+			return nil, Cursor{}, false, err
+		}
+		if !exists {
+			continue
+		}
+		bundleClient := s.bundleManagerClient.BundleClient(batchDumpID)
+
+		results, count, err := bundleClient.MonikerResults(context.Background(), "reference", scheme, identifier, cursor.SkipResultsInDump, limit)
+		if err != nil {
+			return nil, Cursor{}, false, err
+		}
+		if len(results) == 0 {
+			continue
+		}
+		resolvedLocations := resolveLocationsWithDump(dump, results)
+
+		if newResultOffset := cursor.SkipResultsInDump + len(results); newResultOffset < count {
+			newCursor := cursor
+			newCursor.SkipResultsInDump = newResultOffset
+			return resolvedLocations, newCursor, true, nil
+		}
+
+		if i+1 < len(cursor.DumpIDs) {
+			newCursor := cursor
+			newCursor.SkipDumpsInBatch = i + 1
+			newCursor.SkipResultsInDump = 0
+			return resolvedLocations, newCursor, true, nil
+		}
+
+		if cursor.SkipDumpsWhenBatching < cursor.TotalDumpsWhenBatching {
+			newCursor := cursor
+			newCursor.DumpIDs = []int{}
+			newCursor.SkipDumpsInBatch = 0
+			newCursor.SkipResultsInDump = 0
+			return resolvedLocations, newCursor, true, nil
+		}
+
+		return resolvedLocations, Cursor{}, false, nil
+	}
+
+	return nil, Cursor{}, false, nil
+}
+
+func applyBloomFilter(refs []db.Reference, identifier string, limit int) ([]db.Reference, int) {
+	var filteredReferences []db.Reference
+	for i, ref := range refs {
+		test, err := decodeAndTestFilter([]byte(ref.Filter), identifier)
+		if err != nil || !test {
+			continue
+		}
+
+		filteredReferences = append(filteredReferences, ref)
+
+		if len(filteredReferences) >= limit {
```

Comment:
```suggestion
		if len(filteredReferences) > limit {
```
if my above guess was correct :-)

Reply:
This change fails a test, so it looks like ti's no good!

Thanks for the eyes though! Made me really check that these conditions were correct.